### PR TITLE
Clean up handling of special files

### PR DESCRIPTION
**Pull request by akalin-keybase**
Try to make libdokan and libfuse handling
of special files closer to one another.

Check for zero FolderBranches and TlfIDs
and panic in that case (which may happen
when a TLF hasn't been created yet).

Fixup favorite adding code slightly to
avoid spurious error messages.

### CODE REVIEW

File: libdokan/fs.go [hunk 1]
```diff
@@ -263,14 +263,22 @@ func (f *FS) open(ctx context.Context, oc *openContext, ps []string) (dokan.File
...
 		return NewMetricsFile(f), false, nil
-	case libfs.StatusFileName == ps[0]:
-		return NewStatusFile(f.root.private.fs, nil), false, nil
+		// TODO: Port over ProfileList.
 	case libfs.ResetCachesFileName == ps[0]:
 		return &ResetCachesFile{fs: f.root.private.fs}, false, nil
+
+		// This section is equivalent to
+		// handleNonTLFSpecialFile in libfuse.
+	case libfs.StatusFileName == ps[0]:
```
Comment (akalin-keybase) [inline libdokan/fs.go]: @taruti is there a reason why some cases use `ps[psl-1]` and others use `ps[0]`?

Also, `openSpecialFile` (now `handleTLFSpecialFile`) says that common files are handled in this function. But doesn't that mean that this case for `StatusFileName` (which doesn't pass in the TLF) is hit first before the case in `handleTLFSpecialFile`?

File: libkbfs/kbfs_ops.go [hunk 2]
```diff
@@ -185,6 +185,10 @@ func (fs *KBFSOpsStandard) DeleteFavorite(ctx context.Context,
 }
 
 func (fs *KBFSOpsStandard) getOpsNoAdd(fb FolderBranch) *folderBranchOps {
+	if fb == (FolderBranch{}) {
+		panic("zero FolderBranch in getOps")
```
Comment (akalin-keybase) [inline libkbfs/kbfs_ops.go]: I figured this was more fool-proof than trying to plumb the error up from everywhere.

**Comment by akalin-keybase** [comment]
I expect I'll run into libdokan CI failures, unless I magically got it to compile on the first try.

**Comment by strib** [comment]
Looks good to me, thanks, but @taruti should review as well.

File: libdokan/fs.go (hunk 1 above)
Comment (taruti) [inline libdokan/fs.go]: ps[0] means "The first element of the path
ps[psl-1] means "The last element of the path.

e.g. with a path like
`\Private\user1,user\.kbfs_error`
ps[0] = `Private`
ps[psl-1] = `.kbfs_error`

Yes, some special files are handled globally in the fs.go at the moment.

File: libdokan/fs.go [hunk 3]
```diff
@@ -263,14 +263,22 @@ func (f *FS) open(ctx context.Context, oc *openContext, ps []string) (dokan.File
...
+
+		// This section is equivalent to
+		// handleCommonSpecialFile in libfuse.
 	case libkbfs.ErrorFile == ps[psl-1]:
 		return NewErrorFile(f), false, nil
 	case libfs.MetricsFileName == ps[psl-1]:
 		return NewMetricsFile(f), false, nil
-	case libfs.StatusFileName == ps[0]:
-		return NewStatusFile(f.root.private.fs, nil), false, nil
+		// TODO: Port over ProfileList.
```
Comment (taruti) [inline libdokan/fs.go]: ProfileList is ported over some 20 lines below this - feel free to move it here.

Comment (taruti) [inline libdokan/fs.go]: However do remember that when returning from file open we return 3 params - 
one of which is the boolean "is it a directory" flag.

**Comment by taruti** [comment]
One thing to think about is whether we want a helper for SpecialReadFiles that lack a time.Time,
currently we have boilerplate like:
`
bs, err := foo
return bs, time.Time{}, err
`
in multiple places.

The Dokan specific parts look good modulo comments.

File: libdokan/fs.go (hunk 1 above)
Comment (akalin-keybase) [inline libdokan/fs.go]: Yeah, that's what I thought, but does that mean that e.g. `\.kbfs_status\foo\bar` will trigger the status file case? shouldn't we condition on `psl == 1` also?

Comment (taruti) [inline libdokan/fs.go]: Yes, but that is harmless. Feel free to add those if you want.

**Comment by akalin-keybase** [comment]
Filed KBFS-1498 and KBFS-1499 for CI failures, which I believe are unrelated.

File: libdokan/fs.go (hunk 3 above)
Comment (akalin-keybase) [inline libdokan/fs.go]: Ah, thanks, rearranged to match order with libfuse.

File: libdokan/fs.go (hunk 1 above)
Comment (akalin-keybase) [inline libdokan/fs.go]: Hmm...given the trouble I had getting tests to pass for this one, I'll leave that for a future PR. :)

**Comment by akalin-keybase** [comment]
Yeah, SpecialReadFile should probably take a fleshed-out interface instead of a function pointer. And that interface could be nice enough to avoid having to drop the returned time, perhaps.

**Comment by akalin-keybase** [comment]
See https://keybase.atlassian.net/browse/KBFS-1499 for an explanation of the CI failure and why it happened. (I also added comments in the relevant place in the code.)

**Pull request merged**